### PR TITLE
feat: websocket search & result_cache_v2

### PR DESCRIPTION
**Pull request by Loaki07**
# Objectives:
1. Stream results to the client using web-socket.
2. Cache results for a query as a single result cache.

## New Endpoint:
```sh
/api/{org_id}/ws/{req_id}
```

## WS Design:
1. The request_id is used to handle the sessions `pub static WS_SESSIONS: Lazy<RwHashMap<String, WsSession>> = Lazy::new(DashMap::default);`.
2. The request_id is removed from the session cache and the web-socket is disconnected once single search is complete. (e.g. for 5 dashboards in a single page, 5 individual requests are made). The design choice is discussed [here](https://github.com/openobserve/walmart1/issues/207).
3. The supported events and responses the web-socket are described below.
4. Auditing for web-socket linked [here](https://github.com/openobserve/o2-enterprise/pull/248)
5. Org level setting to enable and disable web-socket communication. (Default: disabled)
image

## New Env:
```sh
ZO_WEBSOCKET_ENABLED=true
```

## Env to be enabled:
```sh
ZO_USE_MULTIPLE_RESULT_CACHE=true
ZO_RESULT_CACHE_ENABLED=true
```
- If `ZO_RESULT_CACHE_ENABLED` is set to `false` which is default. Results cache will not be written.

## Nginx config:
```yaml
nginx.ingress.kubernetes.io/proxy-http-version: "1.1" # Enable HTTP/1.1 for WebSockets
nginx.ingress.kubernetes.io/enable-websocket: "true"
nginx.ingress.kubernetes.io/proxy-set-headers: |
        Upgrade $http_upgrade;
        Connection "Upgrade";
```
## Note: Disable `keep-alive` settings if present
```yaml
# nginx.ingress.kubernetes.io/connection-proxy-header: keep-alive
```

## Client Behavior before this pr:
1. For a single query, client calls the `_search_partition` api which returns the list of partitions `Vec<[start_time, end_time]>` -> then makes a `_search` api call for each of the partition.
2. The client also handles the logic for `max_query_range` for the stream.
3. Handle the logic for dashboards loading from right to left irrespective of `order_by` but honors the result by `order_by`

## New Behavior:
1. Client makes a web socket request with a request_id (uuid). This req_id is stored in-memory cache along with its session object.
2. Server Search
    - STEP 1: Search result cache
    - STEP 1(a): Handle the cache responses & query the deltas
    - STEP 2: Search without cache if no caches found
    - STEP 3: Write to result cache
6. Server will now handle querying the partitions and then calling search for each.
7. The results are streamed as in when the responses are received.
8. Server will handle the logic for dashboards loading from right to left irrespective of `order_by` but honors the result by `order_by`

## Modified Code:
1. The `search` service will now use `write_results_v2` when writing to cache.
2. The main difference between `write_results` and `write_results_v2` is the strategy for deduplication. In `write_results_v2` we remove all the hits within the same Date, Hour, Minute and Second; before writing to cache.
```rust
/// # Caching Strategy
/// 1. **Select Hit for Deduplication**:
///    - Selects either the first or last record based on the `is_descending` flag.
///      - `is_descending = true`: Selects the last record.
///      - `is_descending = false`: Selects the first record.
///
/// 2. **Remove All Hits Within the Same Date, Hour, Minute, and Second for Deduplication**:
///    - Identifies the timestamp of the hit to remove (based on the `is_descending` flag, either
///      the first or the last record is chosen).
///    - Removes all hits that match the exact same date, hour, minute, and second
///      (`YYYY-MM-DDTHH:MM:SS`) as the identified hit.
///      - Example: If the timestamp to remove is `2024-12-06T04:15:30`, only hits with the exact
///        timestamp `2024-12-06T04:15:30` are removed, while others with different seconds (e.g.,
///        `2024-12-06T04:15:29` or `2024-12-06T04:15:31`) are retained.
``` 

## Web-socket Events:
1. Search
```sh
{
    "type": "search",
    "content": {
        "trace_id": "d56755eb02cf47ecb3064ee096735c4a",
        "payload": {
            "query": {
                "sql": "select histogram(_timestamp, '1 hour') AS zo_sql_key, count(*) AS zo_sql_num from \"olympics\"  GROUP BY zo_sql_key ORDER BY zo_sql_key",
                "start_time": 1733499294217000,
                "end_time": 1734708894217000,
                "size": -1,
                "query_fn": "LmFiYywgZXJyID0gZ2V0X2VucmljaG1lbnRfdGFibGVfcmVjb3JkKCJvbHltcGljcyIsCnsKICAicmFuayI6IHRvX3N0cmluZyEoLnJhbmspCn0pCi5yYW5rX2tleXdvcmQgPSAuYWJjLmtleXdvcmQKLg..",
                "sql_mode": "full"
            }
        },
        "stream_type": "enrichment_tables",
        "search_type": "ui",
        "use_cache": true
    }
}
```
2. Cancel
```sh
{
    "type": "cancel",
    "content": {
        "trace_id": "b221bb019f3742e8bd93f817d530ef70"
    }
}
```

Related Issues:  #5173, #5188, #5266, #5320

## Release Notes

- **New Features**
	- Introduced WebSocket search functionality with a toggle in settings.
	- Added support for real-time data retrieval via WebSocket in various components.
	- Enhanced search operations with new methods for handling WebSocket queries.
	- Added new properties and methods to improve handling of WebSocket sessions and requests.
	- New interface for search request payloads and improvements in query handling.
	- Implemented WebSocket proxy handling for improved connection management.
	- Enhanced caching functionality with new methods for managing cached responses.
	- Introduced optional field for enabling WebSocket search in organization settings.
	- Added a new constant for handling partial error responses in search operations.
	- Added a new configuration option for enabling WebSocket support.
	- Updated package version and various dependency versions to enhance stability.

- **Bug Fixes**
	- Improved error handling and logging for search requests and caching mechanisms.
	- Enhanced validation and URL construction for alerts.

- **Documentation**
	- Updated comments and documentation for clarity on new features and changes.

- **Style**
	- Minor formatting adjustments for consistency across components and services.

- **Chores**
	- Updated dependency versions to enhance stability and performance.

### CODE REVIEW

File: web/src/components/settings/General.vue [hunk 1]
```diff
@@ -253,6 +266,10 @@ export default defineComponent({
       scrapeIntereval.value =
         store.state?.organizationData?.organizationSettings?.scrape_interval ??
         15;
+
+      enableWebsocketSearch.value =
+        store.state?.organizationData?.organizationSettings
+          ?.enable_websocket_search ?? false;
```
[the remark this answers is not in the document]
Comment (ktx-abhay) [reply web/src/components/settings/General.vue:272]: that is fine. does not required to validate the connection.

**Review by bjp232004 — APPROVED** [review]

**Review by hengfeiyang — APPROVED** [review]

File: web/src/composables/useLogs.ts [hunk 2]
```diff
@@ -4355,6 +4458,800 @@ const useLogs = () => {
...
+          b64EncodeUnicode(searchObj.data.tempFunctionContent) || "";
+      }
+
+      if (searchObj.data.datetime.type === "relative") {
+        if (!isPagination) initialQueryPayload.value = cloneDeep(queryReq);
+        else {
+          if (
+            searchObj.meta.refreshInterval == 0 &&
+            router.currentRoute.value.name == "logs" &&
+            searchObj.data.queryResults.hasOwnProperty("hits")
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4527]: Created a new issue as there are many instances where we need to replace
#5453

File: web/src/composables/useLogs.ts [hunk 3]
```diff
@@ -4355,6 +4458,874 @@ const useLogs = () => {
...
+    } catch (e: any) {
+      console.error("Error while getting data through web socket", e);
+
+      searchObj.loading = false;
+      showErrorNotification(
+        notificationMsg.value || "Error occurred during the search operation.",
+      );
+      notificationMsg.value = "";
+    }
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4602]: Added methods  buildWebSocketPayload and initializeWebSocketConnection.
We have generic method to show error notification showErrorNotification

File: web/src/composables/useLogs.ts [hunk 4]
```diff
@@ -4355,6 +4454,846 @@ const useLogs = () => {
...
+      });
+      // cancelQuery();
+    } catch (e: any) {
+      searchObj.loading = false;
+      showErrorNotification(
+        notificationMsg.value || "Error occurred while sending socket message.",
+      );
+      notificationMsg.value = "";
+    }
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4666]: Done, removed console log.
Also retry mechanism is added on failure

File: web/src/composables/useLogs.ts [hunk 5]
```diff
@@ -4355,6 +4463,848 @@
...
+    searchObj.data.datetime.startTime = extractedDate.from;
+    searchObj.data.datetime.endTime = extractedDate.to;
+  };
+
+  // Limit, aggregation, vrl function, pagination, function error and query error
+  const handleSearchResponse = (
+    requestId: string,
+    payload: any,
+    response: any,
+  ) => {
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4679]: Done

File: web/src/composables/useLogs.ts [hunk 6]
```diff
@@ -4355,6 +4458,799 @@ const useLogs = () => {
...
+      ) {
+        searchObj.data.functionError = response.content.results.function_error;
+      }
+
+      if (
+        response.content.results.hasOwnProperty("function_error") &&
+        response.content.results.function_error != "" &&
+        response.content.results.hasOwnProperty("new_start_time") &&
+        response.content.results.hasOwnProperty("new_end_time")
+      ) {
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4731]: Created a new issue as there are many instances where we need to replace
https://github.com/openobserve/openobserve/issues/5453

File: web/src/composables/useLogs.ts [hunk 7]
```diff
@@ -4342,6 +4412,721 @@
...
+          searchObj.data.queryResults.hits[0][
+            store.state.zoConfig.timestamp_column
+          ],
+        );
+        searchObj.data.queryResults.hits = response.content.hits;
+      }
+
+      if (!searchObj.meta.refreshInterval) {
+        // In page count we set track_total_hits
+        if (!queryReq.query.hasOwnProperty("track_total_hits")) {
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4784]: Created a new issue as there are many instances where we need to replace
https://github.com/openobserve/openobserve/issues/5453

File: web/src/composables/useLogs.ts [hunk 8]
```diff
@@ -4355,6 +4458,874 @@
...
+
+      searchObjDebug["paginatedDataReceivedEndTime"] = performance.now();
+    } catch (e: any) {
+      searchObj.loading = false;
+      console.log(e);
+      showErrorNotification(
+        notificationMsg.value || "Error occurred while handling logs response.",
+      );
+      notificationMsg.value = "";
+    }
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4835]: Done

File: web/src/composables/useLogs.ts [hunk 9]
```diff
@@ -4355,6 +4458,874 @@
...
+      errorMsg += traceIdMsg;
+      notificationMsg.value = `${errorMsg} TraceID: ${finalTraceId}`;
+    } else {
+      notificationMsg.value = errorMsg;
+    }
+
+    // Update the error message in the data object
+    searchObj.data.errorMsg = errorMsg;
+    // searchObj.data.errorCode = code || "";
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:5199]: Done

File: web/src/composables/useLogs.ts [hunk 10]
```diff
@@ -4355,6 +4454,846 @@
...
+
+  const handleCancelQuery = () => {
+    $q.notify({
+      message: "Running query cancelled successfully",
+      color: "positive",
+      position: "bottom",
+      timeout: 4000,
+    });
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:5296]: Done

File: web/src/composables/useLogs.ts [hunk 11]
```diff
@@ -4342,6 +4412,721 @@
...
+  const handlePageCountError = (err: any) => {
+    searchObj.loading = false;
+    let trace_id = "";
+    searchObj.data.countErrorMsg = "Error while retrieving total events: ";
+    if (err.response != undefined) {
+      if (err.response.data.hasOwnProperty("trace_id")) {
+        trace_id = err.response.data?.trace_id;
+      }
+    } else {
+      if (err.hasOwnProperty("trace_id")) {
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:5278]: Created a new issue as there are many instances where we need to replace
https://github.com/openobserve/openobserve/issues/5453

File: web/src/composables/useSearchWebSocket.ts [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+      );
+
+      webSocket.addErrorHandler(
+        requestId,
+        handlers.error.bind(null, requestId, data),
+      );
+      return requestId;
+    } catch (e: any) {
+      throw new Error("Error in fetching search data", e);
+    }
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useSearchWebSocket.ts:54]: Done

File: web/src/composables/useSearchWebSocket.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  };
+
+  const sendSearchMessageBasedOnRequestId = (requestId: string, data: any) => {
+    webSocket.sendMessage(requestId, JSON.stringify(data));
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useSearchWebSocket.ts:59]: Done

File: web/src/composables/useWebSocket.ts [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+    socket.addEventListener("open", (event) => onOpen(socketId, event));
+    socket.addEventListener("message", (event) => onMessage(socketId, event));
+    socket.addEventListener("close", (event) =>
+      onClose(socketId, event, interval, maxAttempts, createSocket),
+    );
+  };
+
+  createSocket();
+};
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts:40]: Done

File: web/src/composables/useWebSocket.ts [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+const onMessage = (socketId: string, event: MessageEvent) => {
+  const data = JSON.parse(event.data);
+
+  if (data.type === "error") {
+    errorHandlers[socketId]?.forEach((handler) => handler(data));
+    return;
+  }
+
+  messageHandlers[socketId]?.forEach((handler) => handler(data));
+};
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts:62]: Works fine for now, will add if required

File: web/src/composables/useWebSocket.ts [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+  handlersMap: Record<string, any[]>,
+  socketId: string,
+  handler: any,
+) => {
+  const handlers = handlersMap[socketId];
+  if (handlers) {
+    const index = handlers.indexOf(handler);
+    if (index > -1) handlers.splice(index, 1);
+  }
+};
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts:123]: Done

File: web/src/composables/useWebSocket.ts [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    for (const socketId in sockets) {
+      const socket = sockets[socketId];
+      if (socket) {
+        socket.close();
+        clearInterval(pingIntervals[socketId]);
+        delete sockets[socketId];
+        delete pingIntervals[socketId];
+      }
+    }
+  });
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts:148]: Done

File: web/src/layouts/MainLayout.vue [hunk 18]
```diff
@@ -1080,6 +1080,8 @@ export default defineComponent({
             orgSettings?.data?.data?.trace_id_field_name ?? "traceId",
           toggle_ingestion_logs:
             orgSettings?.data?.data?.toggle_ingestion_logs ?? false,
+          enable_websocket_search:
+            orgSettings?.data?.data?.enable_websocket_search ?? false,
         });
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/layouts/MainLayout.vue:1106]: Done

File: web/src/plugins/logs/Index.vue [hunk 19]
```diff
@@ -369,7 +369,7 @@ import { buildSqlQuery, getFieldsFromQuery } from "@/utils/query/sqlUtils";
 import useNotifications from "@/composables/useNotifications";
 import SearchBar from "@/plugins/logs/SearchBar.vue";
 import SearchHistory from "@/plugins/logs/SearchHistory.vue";
-import type { ActivationState, PageType } from "@/ts/interfaces/logs.ts";
+import { type ActivationState, PageType } from "@/ts/interfaces/logs.ts";
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/plugins/logs/Index.vue:372]: Not required

File: web/src/plugins/logs/Index.vue [hunk 20]
```diff
@@ -369,7 +369,7 @@
...
 
 export default defineComponent({
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/plugins/logs/Index.vue:374]: Will remove and fix in different PR, there are any instances of it

File: web/src/services/search.ts [hunk 21]
```diff
@@ -59,6 +59,44 @@ const search = {
...
+    const payload = {
+      type: "search",
+      content: {
+        trace_id,
+        payload: {
+          query,
+        },
+      },
+    };
+  },
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/services/search.ts:98]: Removed unused method

File: web/src/stores/index.ts [hunk 22]
```diff
@@ -76,6 +76,8 @@ export default createStore({
     savedFunctionDialog: false,
     regionInfo: [],
     hiddenMenus: [],
+    sessionId: "",
+    webSocketUrl: "",
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/stores/index.ts:80]: Not required

File: web/src/stores/index.ts [hunk 23]
```diff
@@ -306,6 +314,12 @@
     setHiddenMenus(context, payload) {
       context.commit("setHiddenMenus", payload);
     },
+    setSessionId(context, payload) {
+      context.commit("setSessionId", payload);
+    },
+    setWebSocketUrl(context, payload) {
+      context.commit("setWebSocketUrl", payload);
+    },
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/stores/index.ts:322]: Removed dead code

File: web/src/composables/dashboard/usePanelDataLoader.ts [hunk 24]
```diff
@@ -74,6 +75,15 @@ export const usePanelDataLoader = (
   };
   let runCount = 0;
 
+  const searchRetriesCount = ref<{ [key: string]: number }>({});
```
[the remark this answers is not in the document]
Comment (ktx-kirtan) [reply web/src/composables/dashboard/usePanelDataLoader.ts:78]: This is added

File: web/src/composables/useLogs.ts [hunk 25]
```diff
@@ -4355,6 +4461,849 @@
...
+    if (customMessage) {
+      errorMsg = t(customMessage);
+    }
+
+    if (trace_id) {
+      errorMsg += ` <br><span class='text-subtitle1'>TraceID: ${trace_id}</span>`;
+    }
+
+    return errorMsg;
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:5190]: Done

File: web/src/composables/useLogs.ts [hunk 26]
```diff
@@ -4355,6 +4461,849 @@
...
+      searchObjDebug["paginatedDataReceivedEndTime"] = performance.now();
+    } catch (e: any) {
+      searchObj.loading = false;
+      console.log(e);
+      showErrorNotification(
+        notificationMsg.value || "Error occurred while handling logs response.",
+      );
+      notificationMsg.value = "";
+    }
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4837]: Created a new issue as there are many instances #5453

File: web/src/composables/useLogs.ts [hunk 27]
```diff
@@ -4355,6 +4461,849 @@ const useLogs = () => {
...
+      addRequestId(requestId, payload.traceId);
+    } catch (e: any) {
+      console.error("Error while getting data through web socket", e);
+      searchObj.loading = false;
+      showErrorNotification(
+        notificationMsg.value || "Error occurred during the search operation.",
+      );
+      notificationMsg.value = "";
+    }
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useLogs.ts:4604]: No need for now.

File: web/src/composables/useWebSocket.ts [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+      socket.addEventListener("close", (event) =>
+        onClose(socketId, event, interval, maxAttempts, createSocket),
+      );
+    } catch (error) {
+      console.error("Error in creating WebSocket", error);
+    }
+  };
+
+  createSocket();
+};
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts:58]: Removed dead code.
Removed reconnection logic from useWebSocket

File: web/src/composables/useWebSocket.ts [hunk 29]
```diff
@@ -0,0 +1,210 @@
...
+const closeSocket = (socketId: string) => {
+  const socket = sockets[socketId];
+  console.log("close socket", socketId, socket.readyState);
+  if (socket && socket.readyState === WebSocket.OPEN) {
+    setTimeout(() => {
+      sendMessage(socketId, JSON.stringify({ type: "close" }));
+      socket.close(1000, "search cancelled");
+    });
+  }
+};
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts]: Outdated

File: web/src/composables/useSearchWebSocket.ts [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+        requestId,
+        handlers.error.bind(null, requestId, data),
+      );
+      return requestId;
+    } catch (error: any) {
+      console.error(
+        `Error in fetching search data: ${error instanceof Error ? error.message : String(error)}`,
+      );
+    }
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useSearchWebSocket.ts]: Not needed

File: web/src/composables/useSearchWebSocket.ts [hunk 31]
```diff
@@ -0,0 +1,93 @@
...
+            trace_id: trace_id,
+          },
+        }),
+      );
+    }
+  };
+
+  const closeSocketBasedOnRequestId = (requestId: string) => {
+    webSocket.closeSocket(requestId);
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useSearchWebSocket.ts:91]: Done

File: web/src/composables/useSearchWebSocket.ts [hunk 32]
```diff
@@ -0,0 +1,105 @@
...
+        handlers.error.bind(null, requestId, data),
+      );
+      return requestId;
+    } catch (error: any) {
+      console.error(
+        `Error in fetching search data: ${error instanceof Error ? error.message : String(error)}`,
+      );
+      return "";
+    }
+  };
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useSearchWebSocket.ts:58]: Works for now

File: web/src/composables/useWebSocket.ts [hunk 33]
```diff
@@ -0,0 +1,196 @@
...
+      socket.addEventListener("open", (event) => onOpen(socketId, event));
+      socket.addEventListener("message", (event) => onMessage(socketId, event));
+      socket.addEventListener("close", (event) => onClose(socketId, event));
+    } catch (error) {
+      console.error("Error in creating WebSocket", error);
+    }
+  };
+
+  createSocket();
+};
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts:51]: Not Required

File: web/src/composables/useWebSocket.ts [hunk 34]
```diff
@@ -0,0 +1,196 @@
...
+  const socket = sockets[socketId];
+  if (socket && socket.readyState === WebSocket.OPEN) {
+    socket.send(message);
+  } else {
+    console.error(
+      `WebSocket ${socketId} is not open. Ready state is:`,
+      socket?.readyState,
+    );
+  }
+};
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/src/composables/useWebSocket.ts:92]: Not required

**Review by bjp232004 — APPROVED** [review]

**Review by ktx-kirtan — APPROVED** [review]

**Review by ktx-abhay — APPROVED** [review]

**Review by hengfeiyang — APPROVED** [review]

**Review by oasisk — APPROVED** [review]

**Pull request merged**